apps/ensnode/src/handlers/Registry.ts

```diff
@@ -124,17 +128,25 @@ export const makeRegistryHandlers = (ownedName: OwnedName) => {
 
         // if the domain doesn't yet have a name, construct it here
         if (!domain.name) {
-          const parent = await context.db.find(schema.domain, { id: node });
+          // attempt to heal the label associated with labelhash via ENSRainbow
+          // https://github.com/ensdomains/ens-subgraph/blob/c68a889/src/ensRegistry.ts#L112-L116
+          const healedLabel = await labelByHash(labelhash);
 
-          // TODO: implement sync rainbow table lookups
-          // https://github.com/ensdomains/ens-subgraph/blob/master/src/ensRegistry.ts#L111
-          const labelName = encodeLabelhash(label);
-          const name = parent?.name ? `${labelName}.${parent.name}` : labelName;
+          // only use the healed label label if it is indexable
+          // otherwise, use encoded labelhash as the label
+          const label = isLabelIndexable(healedLabel) ? healedLabel : encodeLabelhash(labelhash);
+          // update the name using the parent's name and the label value
+          // (either healed label or encoded labelhash)
+          // NOTE: for the root node, the parent is null, so we just use the label value as is
+          const name = parent?.name ? `${label}.${parent.name}` : label;
 
-          await context.db.update(schema.domain, { id: domain.id }).set({ name, labelName });
+          // akin to domain.save() c68a889
+          // https://github.com/ensdomains/ens-subgraph/blob/c68a889e0bcdc6d45033778faef19b3efe3d15fe/src/ensRegistry.ts#L86
+          await context.db.update(schema.domain, { id: domain.id }).set({ name, labelName: label });
```

Comment:
blah, lost my previous comment somehow. tldr, `Domain.labelName` needs to be undefined if label not indexable. current proposed change results in encoded label being stored there. in my suggested change in the other pr i assumed indexable check was done in the healing service client. so updated logic looks something like:

```ts
const healedLabel = await labelByHash(labelhash);
const validLabel = isLabelIndexable(healedLabel) ? healedLabel : undefined;
const label = validLabel || encodeLabelhash(labelhash);
const name = parent?.name ? `${label}.${parent.name}` : label;
await context.db.update(schema.domain, { id: domain.id }).set({ name, labelName: label });
```

Reply:
additional context: [this line](https://github.com/ensdomains/ens-subgraph/blob/c68a889/src/ensRegistry.ts#L113) says "set Domain.labelName to label if it is valid". it doesn't want that label to ever be encoded, which is only used for the domain.name setting logic [here](https://github.com/ensdomains/ens-subgraph/blob/c68a889/src/ensRegistry.ts#L117)